crates/nargo_cli/src/cli/compile_cmd.rs

```diff
@@ -40,10 +47,27 @@ pub(crate) fn run<B: Backend>(
         let compiled_contracts = driver
             .compile_contracts(&args.compile_options)
             .map_err(|_| CliError::CompilationError)?;
-        let preprocessed_contracts = try_vecmap(compiled_contracts, |contract| {
-            preprocess_contract(backend, contract).map_err(CliError::ProofSystemCompilerError)
-        })?;
-        for contract in preprocessed_contracts {
+        // TODO: I wonder if it is incorrect for nargo-core to know anything about contracts.
+        // As can be seen here, It seems like a leaky abstraction where ContractFunctions (essentially CompiledPrograms)
+        // are compiled via nargo-core and then the PreprocessedContract is constructed here.
+        // This is due to EACH function needing it's own CRS, PKey, and VKey from the backend.
+        let preprocessed_contracts: Result<Vec<PreprocessedContract>, CliError<B>> =
+            try_vecmap(compiled_contracts, |contract| {
+                let preprocessed_contract_functions = try_vecmap(contract.functions, |func| {
+                    let common_reference_string =
+                        get_common_reference_string(backend, &func.bytecode)
+                            .map_err(CliError::CommonReferenceStringError)?;
```

Comment:
I'm a little concerned that this will result in us reading huge amounts of data from disk for contracts with many functions. On each function we could be reading and writing 300MB to disk.

Reply:
I think I would split `get_common_reference_string` in two (names are just illustrative):

1. `load_crs_from_disk(backend)` which just grabs the CRS from the disk and doesn't check against the backend to see if we need more.
2. `update_crs_and_cache(backend, circuit) -> Vec<u8>` which will download any extra points we need for `circuit` and then writes the updates crs to file.

This avoids all the unnecessary reads and writes while ensuring that we cache any downloaded extra points. We'll also need a mutable cache variable in this function so that we don't need to redownload the same points for every contract function.

Reply:
I *think* I understand what you are describing. I'll try to implement.

Reply:
Looks good!